enhancements/ingress/wildcard-routing.md

```diff
@@ -0,0 +1,189 @@
+  - "@danehans"
+reviewers:
+  - "@openshift/api-reviewers"
+  - "@openshift/sig-network-edge"
+  - "@danehans"
+
+approvers:
+  - "@openshift/api-approvers"
+creation-date: 2020-03-19
+last-updated: 2020-03-19
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Wildcard Routing Policy
+
+Application developers would like to expose a service outside the cluster by multiple names without specify each one
+individually.
+
+In OpenShift version 4, wildcard routing is not supported. The `Route` resource supports setting a wildcard policy but
+no API exists to enable this feature on an ingress controller (i.e. router).
+
+OpenShift version 3 supports this feature through the `ROUTER_ALLOW_WILDCARD_ROUTES` environment variable when
+configuring the OpenShift Router directly. Users of this feature are blocked from upgrading to OpenShift 4 unless they
+discontinue use of the feature.
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+See the [section][#wildcard-routing-policy]
+
+## Motivation
+
+See the [section][#wildcard-routing-policy]
+
+### Goals
+
+1. Define an API field that allows a cluster administrator to enable wildcard routing for an ingress controller.
+
+### Non-Goals
+
+1. Redesign how wildcard routing is implemented in [OpenShift Router](https://github.com/openshift/router).
+2. Introduce a separate API for managing wildcard routing.
+
+## Proposal
+
+Introduce a `WildcardRouting` field to the [IngressController](https://github.com/openshift/api/blob/master/operator/v1/types_ingress.go)
+resource that allows administrators to enable or disable wildcard routing.
+
+### User Stories
+
+### Implementation Details/Notes/Constraints [optional]
+
+The [Ingress Operator](https://github.com/openshift/cluster-ingress-operator) will use `WildcardRouting` to configure
+the environment variable `ROUTER_ALLOW_WILDCARD_ROUTES` in the [OpenShift Router](https://github.com/openshift/router).
+
+The default behavior should be to disallow wildcard routing, which is backwards compatible with existing
+`IngressController` resources.
+
+### Risks and Mitigations
+
+Addition of the new field requires no migration, as the default is not changing.
+
+## Design Details
+
+### New API Field
+
+`WildcardRouting` enum definition.
+
+```go
+// (Existing fields omitted, only newly proposed fields are visible)
+type RouteAdmissionPolicy struct {
+  // wildcardRouting enables or disables wildcard routing support for
+  // the ingress controller. WildcardRouting works in conjunction with
+  // the WildcardPolicy of routes [1] claimed by the ingress controller.
+  //
+  // [1] https://github.com/openshift/api/blob/master/route/v1/types.go
+  //
+  // WildcardRouting can not be updated.
```

Comment:
Is there a reason why?

Reply:
cc @openshift/sig-network-edge — my inclination is try and support mutation unless there are severe technical roadblocks. One question that comes to mind is what happens to existing routes which were admitted under the previous policy?

Reply:
And by extension, what happens if there are routes previously admitted under one policy and the ingresscontroller is re-created with a new policy?

Reply:
> Is there a reason why?

@ironcladlou I was basing this off the [3.11 docs](https://docs.openshift.com/container-platform/3.11/install_config/router/default_haproxy_router.html#using-wildcard-routes) that show the following:
```
$ oc adm router --replicas=0 ...
$ oc set env dc/router ROUTER_ALLOW_WILDCARD_ROUTES=true
$ oc scale dc/router --replicas=1
```

After giving this more thought, the field can be mutated. Changing the field will cause the router deployment to update due to the env var change.

> One question that comes to mind is what happens to existing routes which were admitted under the previous policy?

I would expect routes that are no longer covered under the policy to not be admitted. Is this acceptable? If not, whenever going from Enabled > Disabled, should the Ingress Operator prevent the `IngressController` change as follows:
- Check if wildcard status is Enabled and spec is Disabled.
- If so, list all routes. For each route with `status.ingress.routerName` that matches the `IngressController` name, check `wildcardPolicy: None`.
- If all routes are `wildcardPolicy: None`, write spec to status for said `IngressController`.
- Otherwise do not write spec to status and surface an `IngressController` status condition.

> And by extension, what happens if there are routes previously admitted under one policy and the ingresscontroller is re-created with a new policy?

I need to look closer at the openshift/router controllers. I would expect a route controller to update route admission status when the associated `IngressController` no longer exists. I just deleted the default ingress controller and console route status continues to indicate `routerName: default`:
```
...
status:
  ingress:
  - conditions:
    - lastTransitionTime: "2020-03-20T16:14:45Z"
      status: "True"
      type: Admitted
    host: console-openshift-console.apps.dhansen.devcluster.openshift.com
    routerCanonicalHostname: apps.dhansen.devcluster.openshift.com
    routerName: default
    wildcardPolicy: None
```
This seems like a bug. Am I missing something?

Reply:
The ingress controller itself (openshift/router) is the thing that manages Route resources.

Reply:
I suspect this behavior is simply undefined right now. Because we'll now effectively enable the option to change it (either through a patch or a replace of the ingresscontroller), seems like we have to define the behavior as part of this EP.

So, forgetting what the undefined behavior happens to be today, what _should_ happen to existing routes that were admitted under the previous policy? They could:

1. Remain admitted even though they're now technically invalid — if they are subsequently deleted and re-created, they would be rejected, but would continue to work while in their invalid state.
2. Be retroactively rejected, breaking the newly invalidated routes.
3. ???

cc @openshift/sig-network-edge @smarterclayton

Reply:
@ironcladlou I updated the EP as such:

> Updating an ingress controller from WildcardsAllowed to WildcardsDisallowed will cause previously admitted routes with `wildcardPolicy: Subdomain` to no longer work even though status indicates otherwise. These routes can not be updated since `wildcardPolicy` is immutable. Instead, these routes must be deleted and created with `wildcardPolicy: None`.

Reply:
What about something like:

>Changing an existing ingress controller's `wildcardPolicy` _may_ cause disruption to routes already admitted by the ingress controller.
>
>Any admitted routes which are retroactively invalidated by a change to `wildcardPolicy`  may experience a disruption.

Regarding route status becoming out of alignment with reality, what if we treat that as a bug in the implementation (the system should be smart enough to re-admit routes and fix up the status on its own).

Reply:
@ironcladlou I updated the wording to draw greater attention to this caveat and included the same update in the associated openshift/api PR. @Miciah and I discussed this issue as an openshift/router bug yesterday: https://coreos.slack.com/archives/CCH60A77E/p1585261768074200. I'll file a bug for this issue. Should a reference to the bug be included in the godocs or simply be ref'd in this PR?

Reply:
@ironcladlou I like that my docs explicitly spell out the issue. For example, changing an ingress controller's wildcardPolicy from wildcardsDisallowed to WildcardsAllowed will cause no disruption to previously admitted routes.

Reply:
SGTM — as a followup we should figure out how to let users know about the newly broken routes and how they should deal with them, perhaps driven top-down from operator status